### PR TITLE
core/rust: T1 UI support

### PR DESCRIPTION
**Pull request by mmilata**
Adds basic support for T1 layouts written in rust. The aim of this PR is to make it possible to develop the T1 and TT UI mostly independently. The Button & Dialog implementation will likely be rewritten (input event handling needs to happen in Dialog to support pressing both buttons) and the confirm_action layout is also extremely rudimentary.

### CODE REVIEW

File: core/embed/rust/src/ui/component/model.rs
Comment (mmilata) [inline core/embed/rust/src/ui/component/model.rs:67]: Not sure whether making this a type parameter of `struct Event` wouldn't be cleaner than feature-dependent type alias. I wanted to avoid additional type parameters when there is only one possibility in given build, but don't really know whether it's the Right Thing.

File: core/embed/rust/src/micropython/func.rs
Comment (jpochyla) [inline core/embed/rust/src/micropython/func.rs:41]: This definitely needs to be `unsafe fn`.

File: core/embed/rust/src/ui/component/model.rs
Comment (jpochyla) [reply core/embed/rust/src/ui/component/model.rs:67]: I'd avoid the type alias.

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (jpochyla) [inline core/embed/rust/src/ui/component/model_t1/button.rs:131]: Personally i'd prefer to have the layout (bounding box) calculations outside this function.

File: core/embed/rust/src/ui/component/model_t1/dialog.rs
Comment (jpochyla) [inline core/embed/rust/src/ui/component/model_t1/dialog.rs:21]: I know using generics in this case is a bit unfortunate, but using `Buffer` here couples the UI to the MicroPython module. This might be OK, I don't know, but I've tried to avoid it. What do you think?

File: core/embed/rust/src/ui/component/model_t1/dialog.rs
Comment (jpochyla) [inline core/embed/rust/src/ui/component/model_t1/dialog.rs:36]: Similarly to the comment over `Button::paint`, could we move this magic numbers somewhere?

File: core/embed/rust/src/ui/display.rs
Comment (jpochyla) [inline core/embed/rust/src/ui/display.rs:64]: I think this is fine as it is for now, but in the future i'd really like to expose a more low-level interface to the display for both models. Something like `draw_iter(r: Rect, it: impl IntoIterator<Item = Color>)`.

File: core/embed/rust/src/ui/layout/confirm_t1.rs
Comment (jpochyla) [inline core/embed/rust/src/ui/layout/confirm_t1.rs:35]: Could we use kwargs here?

**Review by jpochyla — COMMENTED** [review]
Nice! I left a few comments.

File: core/embed/rust/src/micropython/func.rs
Comment (mmilata) [reply core/embed/rust/src/micropython/func.rs]: Good point. I've removed it entirely in db922be161c0da87925763260c351983fe676f52, as the helpers in `src/util.rs` already do this and more.

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/button.rs:168]: Yeah the layout calculations were trash, I rewrote them in c7b68d269d9641146d843a7d2035a1ffdf500edc, now they're computed top-down like in model T.

File: core/embed/rust/src/ui/component/model_t1/dialog.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/dialog.rs]: Agreed that we shouldn't be tied to MicroPython unless needed. The `AsRef<[u8]>` solution is actually pretty nice (I just need to figure out some better var name than U). Fixed in c7b68d269d9641146d843a7d2035a1ffdf500edc.

File: core/embed/rust/src/ui/layout/confirm_t1.rs
Comment (mmilata) [reply core/embed/rust/src/ui/layout/confirm_t1.rs]: Good idea, I converted this to kwargs in ef1b60347a3701ee3e2a1c85d6255fa3ef7c98d2 and c7b68d269d9641146d843a7d2035a1ffdf500edc.

**Review by jpochyla — APPROVED** [review]
Ooo, I like the fixes a lot! Good stuff, thanks!

File: core/embed/rust/src/ui/component/model.rs
Comment (matejcik) [inline core/embed/rust/src/ui/component/model.rs:30]: would be nice if we could share the constants between C, Rust, and Micropython.

but they're `#define`s so bindgen won't import them, right?
wondering if it's worth exporting them as consts.

**Review by matejcik — CHANGES_REQUESTED** [review]
Great work overall, this is shaping up nicely.
Some comments below.

High-level comment: IMHO `model -> {components, layouts}` is a better hierarchy here than `components -> model` and `layouts -> model`, and forces us to use only the appropriate components in a given model. Same applies for what is now a `HidEvent` alias -- the layouts should really be using the specific type (`TouchEvent`, `ButtonEvent` maybe instead of putting model name there?) and not the aliased one.

I think we can keep the hierarchy and the event aliases as-is for the purposes of this PR, but we should make a subsequent one that rearranges this.

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/button.rs:86]: thinking about it T1 design has no disabled buttons so we can probably remove it altogether

File: core/embed/rust/src/ui/display.rs
Comment (mmilata) [reply core/embed/rust/src/ui/display.rs:78]: yeah, possibly, though it shouldn't ever make it to the fw image if it's never used

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (jpochyla) [reply core/embed/rust/src/ui/component/model_t1/button.rs:86]: I think helper methods like this help a bit with some logic (like the one in keyboards).

File: core/embed/rust/src/ui/component/model.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model.rs:30]: was also thinking about how to do this, would be nice for e.g. display sizes which are now defined separately too

File: core/embed/rust/src/ui/component/model_t1/dialog.rs
Comment (jpochyla) [reply core/embed/rust/src/ui/component/model_t1/dialog.rs:17]: It's idiomatic to keep the parameters short (and obviously distinct from other names). I'm not fighting to death for this, but you'll see it a lot around Rust code.

File: core/embed/rust/src/ui/component/model.rs
Comment (jpochyla) [reply core/embed/rust/src/ui/component/model.rs:30]: > but they're #defines so bindgen won't import them, right?

I think it should.

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (matejcik) [reply core/embed/rust/src/ui/component/model_t1/button.rs:86]: It's not free in programmer cognitive load though. Not opposed to shorthands that do very similar but slightly different things, but in this case it's literally one `!` away from the other method. As a user of the API, it would make me stop and think if there's some subtle reason to use one or the other.

Comment (jpochyla) [reply core/embed/rust/src/ui/component/model_t1/button.rs:86]: I have an exactly opposite experience 🤷

File: core/embed/rust/src/ui/component/model_t1/dialog.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/dialog.rs:17]: so perhaps `TH, TC, TB`?

Comment (matejcik) [reply core/embed/rust/src/ui/component/model_t1/dialog.rs:17]: :woman_shrugging: My issue is that I don't know at a glance what these types are. I don't know with `TH, TC, TB` either. If this _is_ idiomatic, perhaps just an explanatory comment?

File: ci/test.yml
Comment (mmilata) [reply ci/test.yml]: Fixed in 177221af297b1a80e5ed285cc1acce2f8d6dc66f.

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/button.rs]: Dropped in 6c38a3c397ff37342f00a3b73100bafa3a2a8036.

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/button.rs]: 6c38a3c397ff37342f00a3b73100bafa3a2a8036

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/button.rs]: I've cleaned up the T1 button handling a bit in 6c38a3c397ff37342f00a3b73100bafa3a2a8036, thx for the suggestions.

File: core/embed/rust/src/ui/component/model_t1/button.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/button.rs]: Removed `background_color` in 6c38a3c397ff37342f00a3b73100bafa3a2a8036.

File: core/embed/rust/src/ui/layout/confirm_t1.rs
Comment (mmilata) [reply core/embed/rust/src/ui/layout/confirm_t1.rs]: Agree, the paragraph API proved useful in python, and I think it deserves its own PR.

File: core/embed/rust/src/ui/component/model_t1/dialog.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/dialog.rs]: Fixed in 6c38a3c397ff37342f00a3b73100bafa3a2a8036.

File: core/embed/rust/src/ui/component/model_t1/dialog.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model_t1/dialog.rs]: Fixed in 6c38a3c397ff37342f00a3b73100bafa3a2a8036.

File: core/embed/rust/src/ui/component/model.rs
Comment (mmilata) [reply core/embed/rust/src/ui/component/model.rs]: I've verified that it is possible to import #defines with bindgen. However this change would IMO require some refactoring of the C code so I'd like to leave it to another PR.

**Comment by mmilata** [comment]
I've rearranged the hiearchy in 48cac1f2a324863e77aed158b7607757eb89f531 and 26477d7966127ef9febd07e52ee661d831c27391 so now it's possible to compile emulator with both `model_t1` and `model_tt` enabled at the same time. This causes `#[cfg(feature = ...)]` to infect bunch of other places. Also had to rework how default text formatting is determined and probably missed some simpler solution. Please take a look!

File: core/embed/rust/src/ui/layout/confirm_t1.rs
Comment (jpochyla) [reply core/embed/rust/src/ui/layout/confirm_t1.rs]: Yes, dynamically constructing format strings is a bit of an anti-pattern. I'll prepare a bit nicer API based on top of `Op` iterator that are already there (avoiding format string parser altogether).

**Review by matejcik — APPROVED** [review]
LGTM, i like the reorg.

This should not affect anything, but let's merge _after_ the current freeze anyway.

**Comment by matejcik** [comment]
oh, one thing: the unit tests are failing hard

**Comment by mmilata** [comment]
Rebased, tests pass, will merge after freeze.

**Pull request merged**
